### PR TITLE
ENG-16609, stop deleting empty pbd segments in the initialization of recovery

### PR DESCRIPTION
**Pull request by luyangco**
Delete empty segment in the middle of pbd segment chain would fail the integrity
check in next recovery.

Change-Id: I542e874bae6590a92f3c0e37a8c47fcd635ff7d7

### CODE REVIEW

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 1]
```diff
@@ -625,36 +625,16 @@ private void recoverSegment(long segmentIndex, long segmentId, PbdSegmentName se
...
-                m_segments.put(segment.segmentIndex(), segment);
-            } catch (IOException e) {
-                m_usageSpecificLog.warn(
-                        "Failed to retrieve entry count from segment " + segment.file() + ". Quarantining segment", e);
-                quarantineSegment(segment);
-                return;
-            } finally {
-                segment.close();
+            // Any recovered segment that is not final should be checked for internal consistency.
+            if (!segment.isFinal()) {
```
Comment (rdykiel) [inline src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Why was the try/catch block removed? isFinal() can throw an exception and it was caught here to ensure the segment would be properly closed and quarantined.

**Comment by luyangco** [comment]
> Is this a fix by code inspection of were you able to positively identify the root cause on a bug repro? I would like to understand in which case we can have an empty segment between 2 non-empty segments. Is this a case where the initialization sequence creates an empty segment for inserting new buffers and immediately closes it, and reopens another one for insertion? I think I came across this behavior when testing. If this is the case then the initialization sequence should be fixed to avoid this.

This is a bug found by the system test, please refer https://issues.voltdb.com/browse/ENG-16609.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java (hunk 1 above)
Comment (luyangco) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: isFinal() is just a getter function, it returns the m_final flag or always returns false for quarantined segment.

**Comment by rdykiel** [comment]
Right the bug was found by the system test, but my question was about how the root cause was identified, i.e. did we repro the bug with debug logging showing the creation and deletion of segment files. Also the reply does not address my question about how we can have an empty segment between 2 non-empty segments. It looks like we're fixing the symptom (the empty segment), rather than the cause (why it got created). Finally, did we verify that having an empty segment will not cause issues down the line (e.g. in the truncation logic, etc...)

Also, I think it might help to have a quick summary of the root cause written in the Jira ticket.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java (hunk 1 above)
Comment (rdykiel) [reply src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Yes you're right sorry I should drink more coffee in the morning

**Comment by luyangco** [comment]
Sorry for the confusion, it looks like you missed my last comment in Jira ticket. The direct root cause is that empty segment between two non-empty segments is deleted, as you can see in my last Jira comment. The most likely reason for the creation of empty segment in the middle would be two consecutive UACs without any export data write to stream, each one creates a new segment with empty entry. Segment truncation won't cause this issue because it is always delete segments from the start or the end, the naming chain of segments is still continuous.

File: src/frontend/org/voltdb/utils/PersistentBinaryDeque.java [hunk 2]
```diff
@@ -624,9 +624,9 @@ private void recoverSegment(long segmentIndex, long segmentId, PbdSegmentName se
         } else {
             segment = m_pbdSegmentFactory.create(segmentIndex, segmentId, segmentName.m_file, m_usageSpecificLog,
                     m_extraHeaderSerializer);
-
```
Comment (n3world) [inline src/frontend/org/voltdb/utils/PersistentBinaryDeque.java]: Minor: I would keep the empty line to visually separate the assignment from the work logic

**Pull request merged**